runtime/src/main/java/org/corfudb/util/concurrent/SingletonResource.java

```diff
@@ -57,7 +56,17 @@ private SingletonResource(Supplier<T> generator) {
      * @return The resource provided by this {@link SingletonResource}.
      */
     public T get() {
-        return resource.updateAndGet(t -> t == null ? generator.get() : t);
+        T temp = resource;
+        if (temp == null) {
+            synchronized (this) {
+                temp = resource;
+                if (temp == null) {
+                    temp = generator.get();
+                    resource = temp;
+                }
+            }
+        }
+        return temp;
```

Comment:
There is a possibility of a race condition in which get() returning an instance where resource is already cleaned up.

Reply:
How is that different from calling get from one thread and then calling destroy from another on the same object?

Having cleanup as a part of the singleton seems wrong, it breaks the semantics of a singleton (i.e. recycling it with a new object). The resource should be cleaned up externally. In other words, acquire the resource and then call clean up on it.

Reply:
The race condition happens in the context you mentioned. However temp still holding a reference to the object even if cleaner has nulled or cleaned the resource. So even though the singleton class's resource is null. Get returns its ref to the object through temp. What about returning resource? (it's volatile so either going through initialization or not would be fine).
To you second point, I agree that cleanup breaks the semantic of singleton as we know it.

Reply:
Removed the cleanup logic.